pandajedi/jedidog/AtlasTaskWithholderWatchDog.py

```diff
@@ -0,0 +1,164 @@
+            allSiteList.append(siteName)
+        self.allSiteList = allSiteList
+
+    # get map of site to list of RSEs
+    def get_site_rse_map(self, prod_source_label, job_label):
+        site_rse_map = {}
+        for tmpPseudoSiteName in self.allSiteList:
+            tmpSiteSpec = self.siteMapper.getSite(tmpPseudoSiteName)
+            tmpSiteName = tmpSiteSpec.get_unified_name()
+            scope_input, scope_output = DataServiceUtils.select_scope(tmpSiteSpec, prod_source_label, job_label)
+            endpoint_token_map = tmpSiteSpec.ddm_endpoints_input[scope_input].getTokenMap('input')
+            # fill
+            site_rse_map[tmpSiteName] = list(endpoint_token_map.values())
+        # return
+        return site_rse_map
+
+    # get busy sites
+    def get_busy_sites(self, gshare):
+        busy_sites_list = []
+        for tmpPseudoSiteName in self.allSiteList:
+            tmpSiteSpec = self.siteMapper.getSite(tmpPseudoSiteName)
+            tmpSiteName = tmpSiteSpec.get_unified_name()
+            # get nQueue and nRunning
+            nRunning = AtlasBrokerUtils.getNumJobs(jobStatPrioMap, tmpSiteName, 'running', gshare)
+            nQueue = 0
+            for jobStatus in ['defined', 'assigned', 'activated', 'starting']:
+                nQueue += AtlasBrokerUtils.getNumJobs(jobStatPrioMap, tmpSiteName, jobStatus, gshare)
+            # busy sites
+            if (nQueue + 1)/(nRunning + 1) > 2:
+                busy_sites_list.append(tmpSiteName)
+        # return
+        return busy_sites_list
+
+    # # handle waiting jobs
+    # def do_make_tasks_pending(self, task_list):
+    #     tmpLog = MsgWrapper(logger, 'do_make_tasks_pending')
+    #     tmpLog.debug('start')
+    #     # check every x min
+    #     checkInterval = 20
+    #     # make task pending
+    #     for taskID, pending_reason in task_list:
+    #         tmpLog = MsgWrapper(logger, '< #ATM #KV do_make_tasks_pending jediTaskID={0}>'.format(taskID))
+    #         retVal = self.taskBufferIF.makeTaskPending_JEDI(taskID, reason=pending_reason)
+    #         tmpLog.debug('done with {0}'.format(retVal))
+
+
+    # set tasks to be pending due to condition of data locality
+    def do_for_data_locality(self):
+        # refresh
+        self.refresh()
+        # list of job label, gshare, resource type
+        job_label_list = JobUtils.job_labels
+        gshare_list = [ res['name'] for res in self.taskBufferIF.getGShareStatus() ]
+        resource_type_list = self.taskBufferIF.load_resource_types()
+        # parameter from GDP config
+        upplimit_ioIntensity = self.taskBufferIF.getConfigValue('????', '??????', 'jedi', self.vo)
+        lowlimit_currentPriority = self.taskBufferIF.getConfigValue('????', '??????', 'jedi', self.vo)
+        # loop
+        grand_iter = itertools.product(self.prodSourceLabelList, job_label_list, gshare_list, resource_type_list)
+        for prod_source_label, job_label, gshare, resource_type in grand_iter:
```

Comment:
Is this loop OK?

Do we add new gdpconfig parameters for pending tasks? Or are they already some available gdpconfig values for this?

Reply:
You don't have to loop over job_labels which is essentially equivalent to prodSourceLabelList. Basically job_label is used only to flag jobs and most components in JEDI are partitioned by prodsourcelabel. Also you need to get gshare_list for each prodsourcelabel. For example, prodsourcelabel=user doesn't have MC16* gshares.

> Do we add new gdpconfig parameters for pending tasks?

Yes. Perhaps params for each prodsourcelabel as this is not be used for user tasks for now.